File: smritisetu-ai/services/ml-service/app/rules/engagement.py
"""
Which activity to offer next.

This is an engagement suggestion only — it is about how the patient's next few
minutes should feel, not about their abilities.
"""
from __future__ import annotations

from collections import Counter
from datetime import timedelta

from ..config import ENGAGEMENT_ENGINE_VERSION, settings
from ..safety import assert_safe_text
from ..schemas import EngagementRequest, EngagementResponse, SessionMetric

ALL_ACTIVITIES = ["MEMORY_MATCH", "WHO_IS_THIS", "ROUTINE_BUILDER", "MEMORY_LANE"]
CALM_ACTIVITY = "MEMORY_LANE"
RECOGNITION_ACTIVITY = "WHO_IS_THIS"


def recommend(request: EngagementRequest) -> EngagementResponse:
    sessions = sorted(request.sessions, key=lambda s: s.played_at, reverse=True)

    if not sessions:
        return EngagementResponse(
            recommended_game_type="MEMORY_MATCH",
            reason_code="NO_HISTORY",
            explanation=assert_safe_text(
                "No activities have been recorded yet, so a gentle picture-matching activity is offered first."
            ),
            model_version=ENGAGEMENT_ENGINE_VERSION,
        )

    recent = sessions[: settings.abandonment_window]
    abandoned = sum(1 for s in recent if s.abandoned)
    if abandoned >= settings.abandonment_threshold:
        return EngagementResponse(
            recommended_game_type=CALM_ACTIVITY,
            reason_code="RECENT_ABANDONMENT",
            explanation=assert_safe_text(
                f"{abandoned} of the last {len(recent)} activities were left before finishing, "
                "so a calm reminiscence activity with no score is offered next."
            ),
            model_version=ENGAGEMENT_ENGINE_VERSION,
        )

    # A run of effortful sessions -> offer recognition, which is warmer.
    effortful = [s for s in recent if s.accuracy is not None and s.accuracy < 0.5]
    if len(effortful) >= 2:
        return EngagementResponse(
            recommended_game_type=RECOGNITION_ACTIVITY,
            reason_code="OFFER_RECOGNITION_ACTIVITY",
            explanation=assert_safe_text(
                "Recent activities have been effortful, so a familiar-faces activity is offered next."
            ),
            model_version=ENGAGEMENT_ENGINE_VERSION,
        )

    # Otherwise rotate towards whatever has been played least recently, so the
    # day has variety.
    played = Counter(s.game_type for s in sessions[:10])
    least_played = min(ALL_ACTIVITIES, key=lambda g: (played.get(g, 0), ALL_ACTIVITIES.index(g)))
    last_played = sessions[0]
    if last_played.game_type == least_played and len(sessions) > 1:
        alternatives = [g for g in ALL_ACTIVITIES if g != least_played]
        least_played = min(alternatives, key=lambda g: played.get(g, 0))

    return EngagementResponse(
        recommended_game_type=least_played,
        reason_code="VARIETY_ROTATION",
        explanation=assert_safe_text(
            "Things are going steadily, so a different activity is offered to keep the day varied."
        ),
        model_version=ENGAGEMENT_ENGINE_VERSION,
    )


def days_since(sessions: list[SessionMetric], game_type: str) -> float | None:
    matching = [s for s in sessions if s.game_type == game_type]
    if not matching:
        return None
    newest = max(matching, key=lambda s: s.played_at)
    return (max(s.played_at for s in sessions) - newest.played_at) / timedelta(days=1)

File: smritisetu-ai/apps/patient-mobile/src/adaptation/engagement.ts
import type { GameType } from "@smritisetu/shared-types";
import { ADAPTATION, ENGAGEMENT_ENGINE_VERSION } from "./thresholds";
import type { AdaptationSession, EngagementDecision } from "./types";

const ALL_ACTIVITIES: GameType[] = ["MEMORY_MATCH", "WHO_IS_THIS", "ROUTINE_BUILDER", "MEMORY_LANE"];
const CALM_ACTIVITY: GameType = "MEMORY_LANE";
const RECOGNITION_ACTIVITY: GameType = "WHO_IS_THIS";

export function recommendEngagement(sessions: AdaptationSession[]): EngagementDecision {
  const ordered = [...sessions].sort((a, b) =>
    a.playedAt < b.playedAt ? 1 : a.playedAt > b.playedAt ? -1 : 0,
  );

  if (ordered.length === 0) {
    return {
      recommendedGameType: "MEMORY_MATCH",
      reasonCode: "NO_HISTORY",
      explanation:
        "No activities have been recorded yet, so a gentle picture-matching activity is offered first.",
      modelVersion: ENGAGEMENT_ENGINE_VERSION,
    };
  }

  const recent = ordered.slice(0, ADAPTATION.abandonmentWindow);
  const abandoned = recent.filter((session) => session.abandoned).length;
  if (abandoned >= ADAPTATION.abandonmentThreshold) {
    return {
      recommendedGameType: CALM_ACTIVITY,
      reasonCode: "RECENT_ABANDONMENT",
      explanation: `${abandoned} of the last ${recent.length} activities were left before finishing, so a calm reminiscence activity with no score is offered next.`,
      modelVersion: ENGAGEMENT_ENGINE_VERSION,
    };
  }

  const effortful = recent.filter((session) => session.accuracy !== null && session.accuracy < 0.5);
  if (effortful.length >= 2) {
    return {
      recommendedGameType: RECOGNITION_ACTIVITY,
      reasonCode: "OFFER_RECOGNITION_ACTIVITY",
      explanation: "Recent activities have been effortful, so a familiar-faces activity is offered next.",
      modelVersion: ENGAGEMENT_ENGINE_VERSION,
    };
  }

  const played = new Map<string, number>();
  for (const session of ordered.slice(0, 10)) {
    played.set(session.gameType, (played.get(session.gameType) ?? 0) + 1);
  }

  let leastPlayed = ALL_ACTIVITIES.reduce((best, game) => {
    const bestCount = played.get(best) ?? 0;
    const nextCount = played.get(game) ?? 0;
    if (nextCount < bestCount) return game;
    if (nextCount === bestCount && ALL_ACTIVITIES.indexOf(game) < ALL_ACTIVITIES.indexOf(best)) {
      return game;
    }
    return best;
  });

  const lastPlayed = ordered[0];
  if (lastPlayed && lastPlayed.gameType === leastPlayed && ordered.length > 1) {
    const alternatives = ALL_ACTIVITIES.filter((game) => game !== leastPlayed);
    leastPlayed = alternatives.reduce((best, game) =>
      (played.get(game) ?? 0) < (played.get(best) ?? 0) ? game : best,
    );
  }

  return {
    recommendedGameType: leastPlayed,
    reasonCode: "VARIETY_ROTATION",
    explanation: "Things are going steadily, so a different activity is offered to keep the day varied.",
    modelVersion: ENGAGEMENT_ENGINE_VERSION,
  };
}
